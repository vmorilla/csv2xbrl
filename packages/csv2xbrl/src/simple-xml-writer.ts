// A very simple XML writer that produces well formatted XBRL documents
// Implemented to replace XMLBuilder2 which does not allow proper XBML formatting

import { mapKeys } from "./minifunctional";

export class XMLWriter {
    private openTags: string[] = [];

    /**
     * 
     * @param output Writeable stream to write the XML document to
     * @param attrs Attributes of the XML declaration
     * @param namespaces An initial list of namespaces by prefix to be used in the document
     */
    constructor(private output: NodeJS.WritableStream,
        attrs: { version?: "1.0", encoding?: string, standalone?: boolean },
        public namespaces: { [prefix: string]: string } = {}) {
        output.write(`<?xml`);
        for (const [key, value] of Object.entries(attrs)) {
            this.output.write(` ${key}="${value}"`);
        }
        this.output.write("?>\n");
    }

    registerNamespace(prefix: string, uri: string): string {
        if (this.openTags.length) {
            throw new Error("Cannot register namespace after writing an element");
        }

        const matchingKeyValues = Object.entries(this.namespaces).find(([key, value]) => value === uri);
        if (matchingKeyValues) {
            return matchingKeyValues[0];
        }

        let newPrefix = prefix;
        let suffix = 0;
        while (newPrefix in this.namespaces) {
            // Prefix is already in use, we need to change it
            newPrefix = `${prefix}${suffix++}`;
        }
        this.namespaces[newPrefix] = uri;

        return newPrefix;
    }

    /**
     * Writes an open element provided the tag and attributes. It should be later be closed by calling the end method.
     * @param tag 
     * @param attrs 
     * @returns The writer object to allow fluent chaining
     */
    elm(tag: string, attrs: { [key: string]: string } = {}) {
        this.openTag(tag, attrs);
        this.output.write(">\n");
        this.openTags.push(tag);
        return this;
    }

    /**
     * Writes a closed element provided the tag, attributes and text content
     * @param tag 
     * @param attrs 
     * @param text 
     * @returns The writer object to allow fluent chaining
     */
    felm(tag: string, attrs: { [key: string]: string } = {}, text?: string) {
        this.openTag(tag, attrs);
        if (text) {
            this.output.write(`>${encodeXML(text)}</${tag}>\n`);
        }
        else {
            this.output.write(` />\n`);
        }
        return this;
    }

    private openTag(tag: string, attrs: { [key: string]: string }) {
        this.indent().write(`<${tag}`);
        // The root element also includes the namespaces attributes
        const nsAttrs = this.openTags.length ? {} : mapKeys(this.namespaces, (prefix) => prefix ? `xmlns:${prefix}` : "xmlns");
        const allAttrs = { ...nsAttrs, ...attrs };
        for (const key of Object.keys(allAttrs).sort()) {
            const value = allAttrs[key];
            this.output.write(this.openTags.length ? ` ${key}="${value}"` : `\n\t${key}="${value}"`);
        }
    }

    /**
     * Writes a text node
     * @param text 
     * @returns The writer object to allow fluent chaining
     */
    txt(text: string) {
        if (!this.openTags.length)
            throw new Error("Cannot write text to document node");

        this.indent().write(encodeXML(text));
        this.output.write('\n');
        return this;
    }

    /**
     * Closes the currently open element
     * @returns The writer object to allow fluent chaining
     */
    end() {
        if (!this.openTags.length)
            throw new Error("Cannot end document node");

        const tag = this.openTags.pop();
        this.indent().write(`</${tag}>\n`);
        return this;
    }

    /**
     * Writes a comment node
     * @param text 
     * @returns The writer object to allow fluent chaining
     */
    comment(text: string) {
        this.indent().write(`<!-- ${encodeXML(text)} -->\n`);
        return this;
    }

    private indent() {
        this.output.write(`${'\t'.repeat(this.openTags.length)}`);
        return this.output;
    }


}

function encodeXML(text: string) {
    function replaceChar(char: string): string {
        switch (char) {
            case '&': return '&amp;';
            case '<': return '&lt;';
            case '>': return '&gt;';
            case '"': return '&quot;';
            case "'": return '&apos;';
            default: throw new Error(`Unexpected character ${char}`);
        }
    }

    return text.replace(/[&<>"']/g, replaceChar);
}