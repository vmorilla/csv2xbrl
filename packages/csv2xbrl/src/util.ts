import { Writable } from "stream";

export class StringWritable extends Writable {
    private _data = '';

    _write(chunk: any, encoding: BufferEncoding, callback: (error?: Error | null) => void): void {
        this._data += chunk.toString();
        callback();
    }

    toString(): string {
        return this._data;
    }
}
