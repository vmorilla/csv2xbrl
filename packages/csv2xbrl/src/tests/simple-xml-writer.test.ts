import xpath from "xpath";
import { XMLWriter } from "../simple-xml-writer";
import { StringWritable } from '../util';
import { DOMParser } from 'xmldom';


const select = xpath.useNamespaces({});

describe(XMLWriter, () => {
    it("should write a simple XML document", () => {
        const output = new StringWritable();
        const writer = new XMLWriter(output, { version: "1.0", encoding: "UTF-8" });
        writer.elm("root").elm("child").txt("text").end().end();
        // Loads the XML document and returns the text content of the specified node.
        const document = new DOMParser().parseFromString(output.toString());

        expect((select("//root/child/text()", document, true) as Node).nodeValue?.trim()).toBe("text");
        expect(output.toString()).toMatchSnapshot();
    });

    it("should write a simple XML document with text node in the same line", () => {
        const output = new StringWritable();
        const writer = new XMLWriter(output, { version: "1.0", encoding: "UTF-8" });
        writer.elm("root").felm("child", {}, "text").end();
        const document = new DOMParser().parseFromString(output.toString());

        expect((select("//root/child/text()", document, true) as Node).nodeValue).toBe("text");
        expect(output.toString()).toMatchSnapshot();
    });

    it("should write a simple XML document with attributes", () => {
        const output = new StringWritable();
        const writer = new XMLWriter(output, { version: "1.0", encoding: "UTF-8" });
        writer.elm("root", { attr1: "value1", attr2: "value2" }).end();
        const document = new DOMParser().parseFromString(output.toString());

        expect((select("//root/@attr1", document, true) as Node).nodeValue).toBe("value1");
        expect((select("//root/@attr2", document, true) as Node).nodeValue).toBe("value2");
        expect(output.toString()).toMatchSnapshot();
    });

    it("should escape special characters", () => {
        const output = new StringWritable();
        const writer = new XMLWriter(output, { version: "1.0", encoding: "UTF-8" });
        writer.felm("root", {}, "a > b");
        const document = new DOMParser().parseFromString(output.toString());

        expect(output.toString().search(/a &gt; b/)).toBeGreaterThan(-1);

        expect((select("//root/text()", document, true) as Node).nodeValue).toBe("a > b");

        expect(output.toString()).toMatchSnapshot();
    });
});