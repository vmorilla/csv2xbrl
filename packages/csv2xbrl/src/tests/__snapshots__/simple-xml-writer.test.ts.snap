// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XMLWriter should escape special characters 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<root>a &gt; b</root>
"
`;

exports[`XMLWriter should write a simple XML document 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<root>
	<child>
		text
	</child>
</root>
"
`;

exports[`XMLWriter should write a simple XML document with attributes 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<root
	attr1="value1"
	attr2="value2">
</root>
"
`;

exports[`XMLWriter should write a simple XML document with text node in the same line 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<root>
	<child>text</child>
</root>
"
`;
